### ISSUE TITLE
Unable to block access to AWS instance metadata service

### ISSUE THREAD

**Issue by dpolitiuk** [projectcalico/calico#4003]
## Expected Behavior

Traffic to AWS instance metadata service (169.254.169.254) is blocked when a policy has been applied
## Current Behavior

The policy is getting applied but access to the IP remains.
## Steps to Reproduce (for bugs)

1. Apply GlobalNetworkPolicy from below to block access to `169.254.169.254`
2. SSH into an existing pod or create a new pod
3. Test access using `curl 169.254.169.254`

## Context

I have been applying the policy below and similar ones to block AWS instance metadata service access (IP 169.254.169.254). In this case, the second IP is getting blocked properly (216.239.36.21, using it just for testing) while 169.254.169.254 is not. I have tried different GlobalNetworkPolicy rules but all of them result in this IP address not getting blocked, while other external IPs are blocked. When I was looking in the node itself I have noticed that this IP has a separate routing table entry, not sure if it related. I am not sure if there is something I am missing as it feels like rules are getting ignored for this internal traffic.

```
---
apiVersion: projectcalico.org/v3
kind: GlobalNetworkPolicy
metadata:
  name: allow-all-traffic-by-default
spec:
  order: 1000
  egress:
  - action: Allow
    destination:
      nets:
        - 0.0.0.0/0
---
apiVersion: projectcalico.org/v3
kind: GlobalNetworkPolicy
metadata:
  name: block-ec2-metadata-egress
spec:
  order: 10
  egress:
  - action: Deny
    protocol: TCP
    destination:
      nets:
      - 169.254.169.254/32
      - 216.239.36.21/32
```

```
root@ip-10-10-3-12:/# route -n
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         10.10.3.1       0.0.0.0         UG    0      0        0 eth0
10.10.3.0       0.0.0.0         255.255.255.0   U     0      0        0 eth0
10.10.3.39      0.0.0.0         255.255.255.255 UH    0      0        0 enic0a79a7ea36
10.10.3.67      0.0.0.0         255.255.255.255 UH    0      0        0 enid1361341711
10.10.3.70      0.0.0.0         255.255.255.255 UH    0      0        0 eni99a3dc9e0d0
10.10.3.72      0.0.0.0         255.255.255.255 UH    0      0        0 eni4d393ea94b6
10.10.3.75      0.0.0.0         255.255.255.255 UH    0      0        0 enifc1e9ffe736
10.10.3.91      0.0.0.0         255.255.255.255 UH    0      0        0 enif9feadaa3f7
10.10.3.96      0.0.0.0         255.255.255.255 UH    0      0        0 eni33b88d26c54
10.10.3.118     0.0.0.0         255.255.255.255 UH    0      0        0 eni5ed86501ab5
10.10.3.142     0.0.0.0         255.255.255.255 UH    0      0        0 eni2f4328f5c8f
10.10.3.174     0.0.0.0         255.255.255.255 UH    0      0        0 enid4d1dfebf3a
10.10.3.217     0.0.0.0         255.255.255.255 UH    0      0        0 eni643a2f58510
10.10.3.225     0.0.0.0         255.255.255.255 UH    0      0        0 eni5d436365907
10.10.3.234     0.0.0.0         255.255.255.255 UH    0      0        0 eni774be7a1552
169.254.169.254 0.0.0.0         255.255.255.255 UH    0      0        0 eth0
```

## Your Environment

* Calico version: 
```
$ calicoctl version
Client Version:    v3.16.1
Git commit:        7426fc44
Cluster Version:   v3.8.1
Cluster Type:      typha,kdd,k8s,ecs
```
* Orchestrator version (e.g. kubernetes, mesos, rkt):
```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-07T21:20:10Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.13-eks-2ba888", GitCommit:"2ba888155c7f8093a1bc06e3336333fbdb27b3da", GitTreeState:"clean", BuildDate:"2020-07-17T18:48:53Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```
* Operating System and version: Testing on latest ubuntu image pod
* Link to your project (optional):

**Comment by dpolitiuk** [comment]
Update:
I have been trying using host endpoints (https://docs.projectcalico.org/reference/resources/hostendpoint) like the one below to block traffic. While it did block IP it was also causing issues with the EKS (we are using managed node group) marking this one as NotReady and causing issues.
```
apiVersion: projectcalico.org/v3
kind: HostEndpoint
metadata:
  name: 10-10-1-206-host-eth0
spec:
  interfaceName: eth0
  node: ip-10-10-1-206.ec2.internal
```
I believe we would need to have a way to block access to the IP only from pods while allowing nodes themselves to communicate with that IP.

**Comment by caseydavenport** [comment]
I think I've seen this work before with a network policy something like this one:

```yaml
kind: GlobalNetworkPolicy
metadata:
  name: block-metadata-service
spec:
  selector: projectcalico.org/orchestrator == "k8s"
  types: ["Egress"]
  order: 500
  egress:
  - action: Deny
    protocol: TCP
    destination:
      ports: [80]
      nets: ["169.254.169.254/32"]
```

Note: I haven't tried this on a cluster myself, so might need some tweaking. It looks pretty similar to the one you posted above though, so surprised it isn't working. 

For the host endpoints, creating an interface-specific host endpoint like the one in your comment will by default deny traffic in / out of that interface, which is likely why the node is going NotReady. Aside from some failsafes that Calico puts in place, you would need to create policies which allow all the necessary node traffic.

**Comment by dpolitiuk** [comment]
After looking more at the issue we found out that the problem was caused by the kube2iam that we have installed on the cluster. From my understanding, kube2iam was adding iptables rule to proxy AWS related traffic and calico restrictions were not applied. We had to edit the kube2iam deployment to restrict access only to namespaces with `iam.amazonaws.com/allowed-roles=...` annotation, which we added to the kube-system namespace to allow pods in there to still access it.
